### PR TITLE
[Clientside Nav] Add more safety around network failures with fallback UI

### PR DESCRIPTION
**Pull request by damassi**
While browsing Sentry I've noticed [occasional errors](https://sentry.io/organizations/artsynet/issues/1564753432/?project=28316&query=is%3Aunresolved+chunk&statsPeriod=24h) appear that look like 

> Loading chunk 24 failed. (error: https://d1s2w0upia4e9w.cloudfront.net/assets/Routes-Works.22820.adcce79ed4f87e3fc6ed.js)

This typically happens when a user (often on a mobile phone) is navigating between routes and their network connection goes out, which leads to a split JS bundle failing to load. 

We've already [added retry support](https://github.com/artsy/force/pull/5186) on the Force side of things, which will re-try the request up to five times. However, that doesn't handle absolute failures, nor does it gracefully degrade with UI that can lead to a recovery.

According to React's [code splitting docs](https://reactjs.org/docs/code-splitting.html), the best way to handle this is within an Error boundary: 

> If the other module fails to load (for example, due to network failure), it will trigger an error. You can handle these errors to show a nice user experience and manage recovery with Error Boundaries. Once you’ve created your Error Boundary, you can use it anywhere above your lazy components to display an error state when there’s a network error.

Since we already have an error boundary at the root of our app we can listen for specific errors here and handle gracefully, prompting the user to reload the page via our standard `ErrorModal`. (In the future we can further flesh this out for other kinds of global errors.)

retry

### CODE REVIEW

File: src/Artsy/Router/Boot.tsx [hunk 1]
```diff
@@ -57,32 +57,30 @@ export class Boot extends React.Component<BootProps> {
...
+        <StateProvider>
+          <Artsy.SystemContextProvider {...contextProps}>
+            <MediaContextProvider onlyMatch={props.onlyMatchMediaQueries}>
+              <ResponsiveProvider
+                mediaQueries={themeProps.mediaQueries}
+                initialMatchingMediaQueries={props.onlyMatchMediaQueries as any}
+              >
+                <Theme>
+                  <Grid fluid>
+                    <ErrorBoundary>
```
Comment (damassi) [inline src/Artsy/Router/Boot.tsx]: Moved the `ErrorBoundary` down so that it starts at our component tree and can inherit styles

File: src/Artsy/Router/__tests__/Boot.test.tsx [hunk 2]
```diff
@@ -56,20 +55,4 @@ describe("Boot", () => {
   it("injects Grid", () => {
     expect(mount(<Boot {...bootProps} />).find("Grid").length).toEqual(1)
   })
-
-  it("catches errors with componentDidCatch", () => {
```
Comment (damassi) [inline src/Artsy/Router/__tests__/Boot.test.tsx]: Moved this test to `ErrorBoundary.test`

File: src/Artsy/Router/__tests__/buildClientApp.test.tsx [hunk 3]
```diff
@@ -6,6 +6,16 @@ import { mount } from "enzyme"
...
 
+jest.mock("Components/NavBar", () => ({
+  NavBar: () => <div />,
+}))
+
+jest.mock("react-relay", () => ({
+  ReactRelayContext: {
+    Provider: ({ children }) => children,
+  },
+}))
```
Comment (damassi) [inline src/Artsy/Router/__tests__/buildClientApp.test.tsx]: The use of `getDerivedStateFromError` revealed that we were masking a test down below with an unrelated error

File: src/Components/Payment/__tests__/PaymentForm.test.tsx [hunk 4]
```diff
@@ -8,6 +8,7 @@ import { creatingCreditCardFailed } from "Apps/Order/Routes/__fixtures__/Mutatio
 jest.mock("react-relay", () => ({
   commitMutation: jest.fn(),
   createFragmentContainer: component => component,
+  createRefetchContainer: component => component,
```
Comment (damassi) [inline src/Components/Payment/__tests__/PaymentForm.test.tsx]: Another subtle test bug that the error boundary caught

File: src/Artsy/Router/Boot.tsx [hunk 5]
```diff
@@ -57,32 +57,32 @@ export class Boot extends React.Component<BootProps> {
     }
 
     return (
-      <ErrorBoundary>
+      <Theme>
```
Comment (zephraph) [inline src/Artsy/Router/Boot.tsx]: Was it necessary to move the existing error boundary in order to get this to work? Are there any tradeoffs in having it as the outermost wrapper vs having it inside the system context provider?

File: src/Components/ErrorBoundary.tsx [hunk 6]
```diff
@@ -18,7 +28,45 @@ export class ErrorBoundary extends React.Component<Props> {
...
+      return {
+        asyncChunkLoadError: true,
+      }
+    }
+  }
+
   render() {
+    if (this.state.asyncChunkLoadError) {
+      return <ErrorModalWithReload show={this.state.asyncChunkLoadError} />
+    }
```
Comment (zephraph) [inline src/Components/ErrorBoundary.tsx]: Are there other cases we should be checking for? It's not clear to me what this component was really doing previously aside from logging things out.

**Comment by zephraph** [comment]
Oh, and one other note. We should clue in design on some of these edge case experiences.

File: src/Artsy/Router/Boot.tsx (hunk 5 above)
Comment (damassi) [reply src/Artsy/Router/Boot.tsx]: This was just so that I could use Theme file inside the error boundary, which shows the nav. 

> Are there any tradeoffs in having it as the outermost wrapper vs having it inside the system context provider

The only trade off i can see is that an error thrown from within the render function of the context provider proper -- not its children -- wouldn't get caught by it, but there isn't too much logic there. We might want to think about it in the future if we decide to expand functionality at that level.

File: src/Components/ErrorBoundary.tsx (hunk 6 above)
Comment (damassi) [reply src/Components/ErrorBoundary.tsx]: Yup, was just used for logging for the time being. I think we can experiment a bit here!

**Comment by damassi** [comment]
> We should clue in design on some of these edge case experiences

Yup, I definitely planned on doing this once we were done with the A/B test.

**Comment by artsyit** [comment]
:rocket: PR was released in `v25.30.0` :rocket:
